<a href="https://colab.research.google.com/github/Yuliana-Lestari/data-science-2026/blob/main/Pertemuan10_YulianaLestari_250401020199.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [47]:
import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import classification_report, roc_auc_score

In [48]:
from google.colab import files

uploaded = files.upload()

Saving telco_churn.csv to telco_churn (3).csv


In [49]:
df = pd.read_csv("telco_churn.csv")

In [50]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [51]:
print(df.shape)

(7043, 21)


In [52]:
print(df["Churn"].value_counts())

Churn
No     5174
Yes    1869
Name: count, dtype: int64


In [53]:
print(df["Churn"].value_counts(normalize=True))

Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64


In [54]:
df = df.drop("customerID", axis=1)

In [55]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

In [56]:
df = df.dropna()

In [57]:
df = pd.get_dummies(df, drop_first=True)

In [58]:
X = df.drop("Churn_Yes", axis=1)

y = df["Churn_Yes"]

In [59]:
X_tr, X_te, y_tr, y_te = train_test_split(

    X,

    y,

    test_size=0.2,

    stratify=y,

    random_state=42

)

In [60]:
rf = RandomForestClassifier(

    n_estimators=300,

    class_weight="balanced",

    random_state=42

)

In [61]:
_ = rf.fit(X_tr, y_tr)

In [62]:
pred = rf.predict(X_te)

In [63]:
proba = rf.predict_proba(X_te)[:,1]

In [64]:
print(classification_report(y_te,pred))

              precision    recall  f1-score   support

       False       0.83      0.90      0.86      1033
        True       0.63      0.49      0.55       374

    accuracy                           0.79      1407
   macro avg       0.73      0.69      0.71      1407
weighted avg       0.78      0.79      0.78      1407



In [65]:
print("ROC-AUC :",roc_auc_score(y_te,proba))

ROC-AUC : 0.8198746188610092


In [66]:
hasil = pd.DataFrame({

    "Probabilitas Churn":proba

})

hasil.head()

,Probabilitas Churn
0,0.010000
1,0.700000
2,0.016667
3,0.045793
4,0.156667


# Kesimpulan

Model Random Forest berhasil digunakan untuk memprediksi customer churn dengan akurasi sebesar **79%** dan nilai **ROC-AUC sebesar 0,820**. Model memiliki kemampuan yang cukup baik dalam membedakan pelanggan yang berpotensi churn dan tidak churn. Meskipun demikian, nilai recall pada kelas churn masih dapat ditingkatkan agar lebih banyak pelanggan berisiko dapat terdeteksi. Hasil probabilitas dari `predict_proba()` dapat dimanfaatkan perusahaan untuk membantu menentukan strategi retensi pelanggan.